File: templating/core/src/main/python/jsonteng/string_resolver.py
# Copyright 2019 VMware, Inc.
# SPDX-License-Indentifier: Apache-2.0

import re

from .exception import InvalidReferenceException, TemplateEngineException, \
    UnresolvableParameterException
from .tags.tag_base import TagBase


class StringResolver(object):
    """
    Resolve a string by parameter expansions.
    """
    def __init__(self, element_resolver, stats):
        """
        Constructs a new string resolver.
        :param element_resolver: Element resolver
        :type element_resolver: 'ElementResolver'
        :param stats: Stats object for stats collection.
        :type stats: 'Stats'
        """
        self._element_resolver = element_resolver
        self._stats = stats

    def resolve(self, str_data, binding_data_list):
        """
        Resolve a string by parameter expansions.
        :param str_data: String to be resolved.
        :type str_data: 'str'
        :param binding_data_list: Binding data list
        :type binding_data_list: 'list'
        :return: JSON values
        """
        str_len = len(str_data)
        stack = list()
        i = 0
        while i < str_len:
            c = str_data[i]
            if c == '\\':
                # skip escaped char
                i += 2
            elif c == '$':
                # find a param start marker "${"
                next_index = i + 1
                if next_index < str_len and str_data[next_index] == '{':
                    # save the param start position
                    stack.append(i)
                    i += 2
                else:
                    i += 1
            elif c == '}':
                i += 1
                if stack:
                    # get param name.
                    param_start = stack.pop()
                    param_name = str_data[param_start+2:i-1]
                    # resolve the param.
                    value = self._resolve_param(param_name, binding_data_list)
                    # update string.
                    sub_str_before_param = str_data[:param_start]
                    sub_str_after_param = str_data[i:] \
                        if i < str_len else ""
                    if sub_str_before_param or sub_str_after_param:
                        # if the value is only part of the original
                        # string, treat it as a string and reprocess
                        new_str = sub_str_before_param + \
                                  str(self._element_resolver.resolve(
                                      value, binding_data_list)) + \
                                  sub_str_after_param
                        str_data = new_str
                        str_len = len(str_data)
                        i = param_start
                    else:
                        # if the value replaces the whole string,
                        # return the value
                        return self._element_resolver.resolve(
                            value, binding_data_list)
            else:
                i += 1
        if stack:
            raise TemplateEngineException(
                'Mis-formed parameterized string "{}".'.format(str_data))
        return str_data

    def _resolve_param(self, param_name, binding_data_list):
        """
        Resove a parameter by returning its value found in binding_data_list.
        :param param_name: Parameter name.
        :type param_name: 'str'
        :param binding_data_list: Binding data list.
        :type binding_data_list: 'list'
        :return: JSON value
        :rtype: JSON object
        """
        separator_indices = \
            StringResolver._collect_separator_indices(param_name)
        for binding_data in binding_data_list:
            try:
                value = StringResolver._find_param(
                    param_name, separator_indices, binding_data)
                self._stats.update_stats(param_name)
                return value
            except InvalidReferenceException:
                pass
        raise UnresolvableParameterException(
            'Unable to resolve parameter "{}".'.format(param_name))

    @staticmethod
    def _find_param(param_name, separator_indices, binding_data):
        """
        Find the parameter value in this bindign_data. The result could be
        None which is a valid JSON value null. The parameter is matched
        as a whole string. If the whole string does not match any, the prefix
        of the first dot is used to sub parameter map and the rest of the
        string is matched against the sub parameter map. This process is
        repeated until a value is found. For example, parameter "x.y.z" is
        matched against the binding_data. If the binding contains "x.y.z",
        its value is returned. Else, "x" is used to match a sub parameter map
        in the binding data. If a sub parameter map is found, "y.z" is used to
        match a parameter in the sub parameter map. This process is repeated.
        :param param_name: Param name string.
        :type param_name: 'str'
        :param separator_indices: A list of parameter separator indices.
        :type separator_indices: 'list'
        :param binding_data: Parameter value map.
        :type binding_data: 'dict'
        :return: JSON value
        """
        count = len(separator_indices) + 1
        token_start = 0
        next_data = binding_data
        for i in range(count):
            if next_data is None or not isinstance(next_data, dict):
                raise InvalidReferenceException("invalid scope")
            key = param_name[token_start:]
            try:
                return StringResolver._match_key(key, next_data)
            except InvalidReferenceException:
                pass
            if i < len(separator_indices):
                token = param_name[token_start:separator_indices[i]]
                next_data = StringResolver._match_key(token, next_data)
                token_start = separator_indices[i] + 1
        raise InvalidReferenceException("mismatch binding data")

    @staticmethod
    def _match_key(key, data):
        """
        Find the key in params. If found, returns value.
        :param key: Key to be matched.
        :type key: 'str'
        :param data: Binding data map to be searched.
        :type data: 'dict'
        :return: JSON value
        """
        m = re.search("([a-zA-Z0-9]+)\[([0-9]+)\]", key)
        if m:
            key = m.group(1)
            index = int(m.group(2))
            if index < 0:
                raise TemplateEngineException(
                    "Parameter index is negative: %s".format(key))
            if key in data:
                value = data[key]
                if isinstance(value, list) and index < len(value):
                    return value[index]
            raise InvalidReferenceException("mismatch binding data")

        if key in data:
            return data[key]
        raise InvalidReferenceException("mismatch binding data")

    @staticmethod
    def _collect_separator_indices(param_name):
        """
        Locate all parameter name separators (dots).
        :param param_name: Parameter name.
        :type param_name: 'str'
        :return: A list of parameter separator indices.
        :rtype: 'list'
        """
        indices = list()
        str_len = len(param_name)
        i = 0
        while i < str_len:
            c = param_name[i]
            if c == '\\':
                i += 2
            elif c == '.':
                indices.append(i)
                i += 1
            else:
                i += 1
        return indices

File: templating/core/src/main/go/src/jsonteng/string_resolver.go
// Copyright 2019 VMware, Inc.
// SPDX-License-Indentifier: Apache-2.0

package jsonteng

import (
	"container/list"
	"encoding/json"
	"fmt"
	"jsonteng/errors"
	intfs "jsonteng/interfaces"
	"regexp"
	"strconv"
)

type stringResolver interface {
	resolveString(strData *string, bindingDataList *list.List) (interface{}, error)
}

type stringResolverImpl struct {
	elementResolver intfs.ElementResolver
	stats           *stats
}

var pattern = regexp.MustCompile("([a-zA-Z0-9]+)\\[([0-9]+)\\]")

func (stringResolver *stringResolverImpl) init(stats *stats) {
	stringResolver.stats = stats
}

func (stringResolver *stringResolverImpl) setElementResolver(elementResolver intfs.ElementResolver) {
	stringResolver.elementResolver = elementResolver
}

func (stringResolver *stringResolverImpl) resolveString(strData *string, bindingDataList *list.List) (interface{}, error) {
	var errl errors.TemplateEngineError
	strLen := len(*strData)
	stack := list.New()
	i := 0
	for i < strLen {
		c := (*strData)[i]
		if c == '\\' {
			i += 2
		} else if c == '$' {
			nextIndex := i + 1
			if nextIndex < strLen && (*strData)[nextIndex] == '{' {
				stack.PushBack(i)
				i += 2
			} else {
				i++
			}
		} else if c == '}' {
			i++
			if stack.Len() > 0 {
				paramStart := stack.Remove(stack.Back()).(int)
				paramName := (*strData)[paramStart+2 : i-1]
				value, err := stringResolver.resolveParam(&paramName, bindingDataList)
				if err != nil {
					return nil, err
				}
				subStrBeforeParam := (*strData)[:paramStart]
				var subStrAfterParam string = ""
				if i < strLen {
					subStrAfterParam = (*strData)[i:]
				}
				if len(subStrBeforeParam) > 0 || len(subStrAfterParam) > 0 {
					substrMid, err := stringResolver.elementResolver.ResolveElement(value, bindingDataList)
					if err != nil {
						return nil, err
					}
					mid, err := dataToString(substrMid)
					if err != nil {
						return nil, err
					}
					newStr := subStrBeforeParam + mid + subStrAfterParam
					strData = &newStr
					strLen = len(*strData)
					i = paramStart
				} else {
					return stringResolver.elementResolver.ResolveElement(value, bindingDataList)
				}
			}
		} else {
			i++
		}
	}
	if stack.Len() > 0 {
		errl = errors.GenericError
		errl.Message = fmt.Sprintf("Mis-formed parameterized string \"%s\".", *strData)
		return nil, &errl
	}
	return *strData, nil
}

func (stringResolver *stringResolverImpl) resolveParam(paramName *string, bindingDataList *list.List) (interface{}, error) {
	separatorIndices := collectSeparatorIndices(paramName)
	for bindingData := bindingDataList.Front(); bindingData != nil; bindingData = bindingData.Next() {
		if value, err := findParam(paramName, separatorIndices, bindingData.Value.(map[string]interface{})); err == nil || err.ErrorCode != errors.INVALID_REFERENCE {
			if value != nil {
				stringResolver.stats.updateStats(paramName)
				return value, nil
			}
		}
	}
	errl := errors.UnresolvableParameter
	errl.Message = fmt.Sprintf("Unable to resolve parameter \"%s\".", *paramName)
	return nil, &errl
}

func findParam(paramName *string, separatorIndices []int, bindingData map[string]interface{}) (interface{}, *errors.TemplateEngineError) {
	tokenStart := 0
	var nextData interface{}
	nextData = bindingData
	var value interface{}
	var err *errors.TemplateEngineError
	for i := 0; i < len(separatorIndices)+1; i++ {
		if _, ok := nextData.(map[string]interface{}); !ok {
			errl := errors.InvalidReference
			errl.Message = fmt.Sprintf("invalid scope")
			return nil, &errl
		}
		key := (*paramName)[tokenStart:]
		if value, err = matchKey(&key, &nextData); err == nil || err.ErrorCode != errors.INVALID_REFERENCE {
			return value, err
		}
		if i < len(separatorIndices) {
			token := (*paramName)[tokenStart:separatorIndices[i]]
			nextData, err = matchKey(&token, &nextData)
			if err != nil && err.ErrorCode != errors.INVALID_REFERENCE {
				return nil, err
			}
			tokenStart = separatorIndices[i] + 1
		}
	}
	errl := errors.InvalidReference
	errl.Message = fmt.Sprintf("mismatch binding data")
	return nil, &errl
}

func matchKey(key *string, data *interface{}) (interface{}, *errors.TemplateEngineError) {
	matches := pattern.FindStringSubmatch(*key)
	if len(matches) == 3 {
		key = &matches[1]
		index, err := strconv.Atoi(matches[2])
		if err != nil || index < 0 {
			errl := errors.GenericError
			errl.Message = fmt.Sprintf("Parameter index is invalid, %s", *key)
			return nil, &errl
		}
		if v, ok := (*data).(map[string]interface{})[*key]; ok {
			if w, ok := v.([]interface{}); ok {
				if index < len(w) {
					return w[index], nil
				}
			}
		}
		errl := errors.InvalidReference
		errl.Message = fmt.Sprintf("mismatch binding data")
		return nil, &errl
	}
	if v, ok := (*data).(map[string]interface{})[*key]; ok {
		return v, nil
	}
	errl := errors.InvalidReference
	errl.Message = fmt.Sprintf("mismatch binding data")
	return nil, &errl
}

func collectSeparatorIndices(paramName *string) []int {
	indices := list.New()
	for i := 0; i < len(*paramName); i++ {
		if (*paramName)[i] == '\\' {
			i++
		} else if (*paramName)[i] == '.' {
			indices.PushBack(i)
		}
	}
	indicesArray := make([]int, indices.Len())
	i := 0
	for index := indices.Front(); index != nil; index = index.Next() {
		indicesArray[i] = index.Value.(int)
		i++
	}
	return indicesArray
}

func dataToString(value interface{}) (string, error) {
	switch v := value.(type) {
	case string:
		return v, nil
	default:
		if s, ok := json.Marshal(v); ok == nil {
			return string(s), nil
		}
		errl := errors.GenericError
		errl.Message = fmt.Sprintf("Json error. %v", v)
		return "", &errl
	}
}
